### ISSUE TITLE
Has anything changed with optimisticResponse in > 3.8

### ISSUE THREAD

**Issue by jnsandrew** [apollographql/apollo-client#11234]
I feel like I'm going crazy here. I have optimisticResponse working on `3.7.17` but doesn't work on `3.8`. I can't see anything obvious in the changelog for that version, but ultimately my UI does not update anymore. Unfortunately I tried reproducing in the template repo but I'm not sure it's possible as it doesn't connect to a live GQL API, I tried replay.io but when I press record the browser just crashes :/

What I'm seeing on 3.8 is the mutation works, I see the correct response from the server, the component re-renders, but the old data is prevalent (the change I expect is toggling `completed` to `true` or `false` for a task item.

```javascript
// Mutation in my button component
const toggleTaskCompleted = async () => {
    const newCompleted = !entity.completed;

    if (updateEntity) {
      await updateEntity({
        variables: {
          id: entity.id,
          data: {
            completed: newCompleted,
          },
        },
        optimisticResponse: {
          updateTaskEntity: {
            __typename: entity.__typename,
            id: entity.id,
            completed: newCompleted,
          },
        },
      });
    }
  };
```
And the actual mutation:

```javascript
const [updateEntity] = useMutation(MUTATION, {
    client: window.entitiesClient, // I do this as I have another client due to multiple GQL APIs
  });
```

Removing `optimisticResponse` from the mutation works but with the slow updating UI obviously. The apollo docs don't seem to have changed and the above should work?

Nothing crazy complicated, but just weird that it doesn't update on > 3.8. Apologies I can't give any more than that for now.

**Comment by phryneas** [comment]
There's no specific change that we made here - but we really need some kind of reproduction here, I'm sorry. 
The simulated link in the reproduction repo should work for that - but you'd have to add an `update` resolver instead of the existing `add` resolver for it to make sense.

**Comment by jnsandrew** [comment]
OK. I will keep digging for now. Discoveries so far:

1. It seems specific to this client (I have another optimisticResponse on the "default" apollo client and it works fine)
2. After the mutation, if I manually console log the cache and to to `data.Task:13` I can see `completed` is set to the correct boolean
3. If I console log `response.data` (i.e. `const {response} = useQuery(...)`) then the `completed` value is old

So it looks like mutation and optimistic response is working, the cache is updated, but the cached query isn't updating

**Comment by jnsandrew** [comment]
Hmm, I have it working by removing the following from my query:

```javascript
    fetchPolicy: 'network-only',
    nextFetchPolicy: 'cache-first',
```

I'm not 100% sure why this is the case between 3.7 => 3.8.

Unless you have anything more to add here @phryneas we can close this issue.

**Comment by phryneas** [comment]
We *did* have a change in the timing when external data comes in once you change the `fetchPolicy` - I think in 3.7.17. Maybe that makes sense of it.
Before, the `fetchPolicy` would be applied immediately after fetching, so if you started a `network-only` request, with a networkPolicy of `cache-first`, cache updates would start coming in immediately after that request started.
Now they would only come in if they happened after that first network request finished.

Does that make sense to you?

**Comment by jnsandrew** [comment]
Your statement on its' own makes sense, though I'm not sure how I can apply that knowledge to my own code 😄 Based on  the above findings I can reproduce the issue I was seeing now:

https://github.com/jnsandrew/apollo-optimistic-reproduce

This is the diff between working and breaking:
https://github.com/jnsandrew/apollo-optimistic-reproduce/commit/a8070d669595e260d0e11282c8ba4c61da6b9105

You can remove `optimisticResponse` on line 73, or the fetchPolicies on line 40-41, and it works.

So I'm just wondering now if this is a bug or intended behaviour?

**Comment by phryneas** [comment]
Thanks for the reproduction!
I can verify that this behaviour still worked in `3.8.0-alpha.14`, but didn't work anymore in `3.8.0-beta.0`.

Time to dig deeper 🧐

**Comment by phryneas** [comment]
Okay, a lot of git bisecting later, the offending commit is \*drumroll\*:

> Custom document transforms (#10509)

I'm sorry @jerelmiller but we're gonna have a very deep look at that one 😢

**Comment by phryneas** [comment]
Yup, found the problem!

[applyNextFetchPolicy](https://github.com/apollographql/apollo-client/blob/18a91184be3e62e354811d35aad0e55c68fae3ca/src/core/ObservableQuery.ts#L662) relies on the reference to the original `options` object!

**Comment by phryneas** [comment]
Could you try the fix from #11235 and give us feedback if it fixes your application?

```sh
npm i @apollo/client@0.0.0-pr-11235-20230921145806
```
